src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -176,10 +183,194 @@ ObjectDataHandler::write_ret do_insertions(
+  }
+  /*
+   * When trying to modify a portion of an object data block, follow
+   * the read-full-extent-then-merge-new-data strategy, if the write
+   * amplification caused by it is not greater than
+   * seastore_obj_data_write_amplification; otherwise, split the
+   * original extent into at most three parts: origin-left, part-to-be-modified
+   * and origin-right.
+   *
+   * all used variable are described as follows:
+   * |<--------------------------------------write_size--------------------------------------->|
+   *                 |<----------------------data_size----------------------->|
+   *                                data_begin         data_end
+   *                                ---------------------------(incoming data)
+   *                 aligned_begin                                  aligned_end
+   *                 ----------------------------------------------------------(incoming data's block-size-aligned range)
+   * |<--left_size-->|                                                        |<--right_size-->|
+   * pin_begin                                                                           pin_end
+   * -------------------------------------(existing extent)   ----------------------------------(existing extent)
+   * left_paddr                                               right_paddr
+   *
+   * pseudocode of splitting:
+   *
+   * if first_paddr/last_paddr is zero:
+   *   prepend/append hole if data_offset/data_end is not aligned,
+   *   reverse zero extent if necessary.
+   *
+   * if first_paddr/last_paddr is relative:
+   *   merge first_pin/last_pin to incoming data
+   *
+   * if first_paddr/last_paddr is absolute, split extent:
+   *   while either left or right is not processed:
+   *     if write/data_size > seastore_obj_data_write_amplification
+   *       select a larger end according to left_size and right_size,
+   *       split selected end.
+   *       write_size -= left_size or right_size
+   *     else
+   *       merge the end that has not been processed
+   */
+  void split() {
+    auto left_size = aligned_begin - pin_begin;
+    auto right_size = pin_end - aligned_end;
+    bool left_set = false;
+    bool right_set = false;
+    if (left_paddr.is_zero()) {
+      write_size -= left_size;
+      left_size = 0;
+      left_set = true;
+    } else if (left_paddr.is_relative() ||
+	       left_paddr.is_delayed()) {
+      data_size += left_size;
+      left_size = 0;
+      left_set = true;
+    }
+    if (right_paddr.is_zero()) {
+      write_size -= right_size;
+      right_size = 0;
+      right_set = true;
+    } else if (right_paddr.is_relative() ||
+	       right_paddr.is_delayed()) {
+      data_size += right_size;
+      right_size = 0;
+      right_set = true;
+    }
+
+    while (!(left_set && right_set)) {
+      if (((double)write_size / (double)data_size) >
+	  crimson::common::get_conf<double>("seastore_obj_data_write_amplification")) {
```

Comment:
if we move decision out of the loop, it may look like:
```cpp
auto limit = crimson::common::get_conf<double>("seastore_obj_data_write_amplification");
if (((double)write_size / (double)data_size) > limit) {
  if (left_size >= right_size && !left_set && left_size != 0) {
    write_size -= left_size;
    left_set = true;
    split_left = true;
  } else if (!right_set && right_size != 0) {
    write_size -= right_size;
    right_set = true;
    split_right = true;
  }
}
if (((double)write_size / (double)data_size) > limit) {
  if (!left_set) {
    left_set = true;
    split_left = true;
  }
  if (!right_set) {
    right_set = true;
    split_right = true;
  }
}
// fallback case, probably it's impossible
if (!left_set) {
  split_left = true;
}
if (!right_set) {
  split_right = true;
}
```
it just unrolls the loop, if you prefer this style, I'll update.

Reply:
Sorry and NVM. Just realized that the write_size and data_size are unchanged in the loop.

Reply:
NVM, I think I have missed the `left_size >= right_size` part, that looks good.